File: GuardianMonsters/battle/src/de/limbusdev/guardianmonsters/battle/ui/widgets/SwitchActiveGuardianWidget.java
/*
 * *************************************************************************************************
 * Copyright (c) 2018. limbusdev (Georg Eckert) - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 * Written by Georg Eckert <georg.eckert@limbusdev.de>
 * *************************************************************************************************
 */

package de.limbusdev.guardianmonsters.battle.ui.widgets;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.scenes.scene2d.Group;
import com.badlogic.gdx.scenes.scene2d.Touchable;
import com.badlogic.gdx.scenes.scene2d.ui.Button;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.ImageButton;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.utils.TextureRegionDrawable;
import com.badlogic.gdx.utils.Align;

import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian;
import de.limbusdev.guardianmonsters.guardians.monsters.Team;
import de.limbusdev.guardianmonsters.services.Services;
import de.limbusdev.guardianmonsters.ui.Constant;
import de.limbusdev.guardianmonsters.ui.widgets.ATeamChoiceWidget;
import de.limbusdev.guardianmonsters.ui.widgets.Callback;
import de.limbusdev.guardianmonsters.ui.widgets.GuardianStatusWidget;
import de.limbusdev.guardianmonsters.ui.widgets.SimpleClickListener;
import de.limbusdev.guardianmonsters.ui.widgets.TeamCircleWidget;
import de.limbusdev.guardianmonsters.ui.widgets.TiledImage;

/**
 * SwitchActiveGuardianWidget
 *
 * @author Georg Eckert 2018
 */

public class SwitchActiveGuardianWidget extends BattleWidget
{
    private GuardianStatusWidget guardianStatusWidget;
    private ATeamChoiceWidget    teamChoiceWidget;
    private Callback backButtonCallback;
    private Callback switchButtonCallback;
    private Image guardianImg;
    private Button switchButton;
    private Button backButton;

    public SwitchActiveGuardianWidget(Skin battleSkin, Skin inventorySkin)
    {
        setSize(Constant.WIDTH, Constant.HEIGHT);
        TiledImage bgImg = new TiledImage(inventorySkin.getDrawable("bg-pattern-3"), 27, 16);
        bgImg.setPosition(0,0, Align.bottomLeft);
        addActor(bgImg);

        guardianStatusWidget = new GuardianStatusWidget(Services.getUI().getInventorySkin());;
        guardianStatusWidget.setPosition(2, Constant.HEIGHT, Align.topLeft);
        addActor(guardianStatusWidget);

        backButton = new ImageButton(inventorySkin, "button-back");
        backButton.setPosition(Constant.WIDTH-2, 2, Align.bottomRight);
        addActor(backButton);

        switchButton = new ImageButton(inventorySkin, "button-switch");
        switchButton.setPosition(2,2,Align.bottomLeft);
        addActor(switchButton);

        teamChoiceWidget = new TeamCircleWidget(Services.getUI().getInventorySkin(), null);
        teamChoiceWidget.setPosition(Constant.WIDTH-2,Constant.HEIGHT/2,Align.right);
        addActor(teamChoiceWidget);

        Group guardianView = new Group();
        guardianView.setSize(140, Constant.HEIGHT-36);
        guardianView.setPosition(Constant.WIDTH/2, Constant.HEIGHT, Align.top);
        Image monsterViewBg = new Image(inventorySkin.getDrawable("menu-col-bg"));
        monsterViewBg.setPosition(2,2,Align.bottomLeft);
        guardianView.addActor(monsterViewBg);
        guardianImg = new Image();
        guardianImg.setSize(128,128);
        guardianImg.setPosition(6,202,Align.topLeft);
        guardianView.addActor(guardianImg);
        addActor(guardianView);
    }

    public void setCallbacks(Callback onBack, Callback onSwitch)
    {
        this.backButtonCallback = onBack;
        this.switchButtonCallback = onSwitch;

        backButton.clearListeners();
        switchButton.clearListeners();
        backButton.addListener(new SimpleClickListener(backButtonCallback));
        switchButton.addListener(new SimpleClickListener(switchButtonCallback));
    }

    public void init(AGuardian guardian, Team team)
    {
        guardianStatusWidget.init(guardian);
        teamChoiceWidget.init(team, team.getPosition(guardian));
        activateGuardian(guardian, team);

        Callback.ButtonID circleMenuCallbacks = nr ->
        {
            activateGuardian(team.get(nr), team);
        };

        teamChoiceWidget.setHandler(circleMenuCallbacks);
    }

    private void activateGuardian(AGuardian guardian, Team team)
    {
        int chosenPos = teamChoiceWidget.getCurrentPosition();
        if(chosenPos >= team.getActiveTeamSize()) {
            switchButton.setTouchable(Touchable.enabled);
            switchButton.setColor(Color.WHITE);
        } else {
            switchButton.setTouchable(Touchable.disabled);
            switchButton.setColor(Color.GRAY);
        }

        guardianStatusWidget.init(guardian);
        int speciesID = guardian.getSpeciesID();
        int currentForm = guardian.getAbilityGraph().getCurrentForm();
        TextureAtlas.AtlasRegion sprite = Services.getMedia().getMonsterSprite(speciesID, currentForm);
        guardianImg.setDrawable(new TextureRegionDrawable(sprite));
    }

    public int getChosenSubstitute()
    {
        return teamChoiceWidget.getCurrentPosition();
    }
}

File: GuardianMonsters/battle/src/de/limbusdev/guardianmonsters/battle/ui/widgets/SwitchActiveGuardianWidget.kt
/*
 * *************************************************************************************************
 * Copyright (c) 2018. limbusdev (Georg Eckert) - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 * Written by Georg Eckert <georg.eckert@limbusdev.de>
 * *************************************************************************************************
 */

package de.limbusdev.guardianmonsters.battle.ui.widgets

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.Group
import com.badlogic.gdx.scenes.scene2d.Touchable
import com.badlogic.gdx.scenes.scene2d.ui.Button
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.ImageButton
import com.badlogic.gdx.scenes.scene2d.utils.Drawable
import com.badlogic.gdx.scenes.scene2d.utils.TextureRegionDrawable
import com.badlogic.gdx.utils.Align

import de.limbusdev.guardianmonsters.guardians.battle.CombatTeam
import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian
import de.limbusdev.guardianmonsters.guardians.monsters.Team
import de.limbusdev.guardianmonsters.scene2d.replaceOnClick
import de.limbusdev.guardianmonsters.services.Services
import de.limbusdev.guardianmonsters.ui.Constant
import de.limbusdev.guardianmonsters.ui.widgets.ATeamChoiceWidget
import de.limbusdev.guardianmonsters.ui.widgets.GuardianStatusWidget
import de.limbusdev.guardianmonsters.ui.widgets.TeamCircleWidget
import de.limbusdev.guardianmonsters.ui.widgets.TiledImage
import ktx.style.get


/**
 * SwitchActiveGuardianWidget
 *
 * @author Georg Eckert 2018
 */

class SwitchActiveGuardianWidget() : BattleWidget()
{
    // .................................................................................. Properties
    private val guardianStatusWidget    : GuardianStatusWidget
    private val teamChoiceWidget        : ATeamChoiceWidget
    private var backButtonCallback      : () -> Unit = {}
    private var switchButtonCallback    : () -> Unit = {}
    private val guardianImg             : Image
    private val switchButton            : Button
    private val backButton              : Button

    val chosenSubstitute: Int
        get() = teamChoiceWidget.currentPosition


    // ................................................................................ Constructors
    init
    {
        val battleSkin = Services.UI().battleSkin
        val inventorySkin = Services.UI().inventorySkin

        setSize(Constant.WIDTHf, Constant.HEIGHTf)
        val bgImg = TiledImage(inventorySkin["bg-pattern-3"], 27, 16)
        bgImg.setPosition(0f, 0f, Align.bottomLeft)
        addActor(bgImg)

        guardianStatusWidget = GuardianStatusWidget()
        guardianStatusWidget.setPosition(2f, Constant.HEIGHTf, Align.topLeft)
        addActor(guardianStatusWidget)

        backButton = ImageButton(inventorySkin, "button-back")
        backButton.setPosition((Constant.WIDTHf - 2), 2f, Align.bottomRight)
        addActor(backButton)

        switchButton = ImageButton(inventorySkin, "button-switch")
        switchButton.setPosition(2f, 2f, Align.bottomLeft)
        addActor(switchButton)

        teamChoiceWidget = TeamCircleWidget(Services.UI().inventorySkin) {}
        teamChoiceWidget.setPosition(Constant.WIDTHf - 2, Constant.HEIGHTf / 2, Align.right)
        addActor(teamChoiceWidget)

        val guardianView = Group()
        guardianView.setSize(140f, Constant.HEIGHTf - 36)
        guardianView.setPosition(Constant.WIDTHf / 2, Constant.HEIGHTf, Align.top)
        val monsterViewBg = Image(inventorySkin.get<Drawable>("menu-col-bg"))
        monsterViewBg.setPosition(2f, 2f, Align.bottomLeft)
        guardianView.addActor(monsterViewBg)
        guardianImg = Image()
        guardianImg.setSize(128f, 128f)
        guardianImg.setPosition(6f, 202f, Align.topLeft)
        guardianView.addActor(guardianImg)
        addActor(guardianView)
    }


    // ..................................................................................... Methods
    fun setCallbacks(onBack: () -> Unit, onSwitch: () -> Unit)
    {
        this.backButtonCallback = onBack
        this.switchButtonCallback = onSwitch

        backButton.replaceOnClick(onBack)
        switchButton.replaceOnClick(onSwitch)
    }

    fun initialize(guardian: AGuardian, team: Team, combatTeam: CombatTeam)
    {
        guardianStatusWidget.initialize(guardian)
        teamChoiceWidget.initialize(team, team.getPosition(guardian))
        activateGuardian(guardian, team, combatTeam)

        val circleMenuCallbacks = { nr: Int -> activateGuardian(team[nr], team, combatTeam) }

        teamChoiceWidget.setHandler(circleMenuCallbacks)
    }

    private fun activateGuardian(guardian: AGuardian, team: Team, combatTeam: CombatTeam)
    {
        if (!combatTeam.isMember(guardian) && guardian.stats.isFit)
        {
            switchButton.touchable = Touchable.enabled
            switchButton.color = Color.WHITE
        }
        else
        {
            switchButton.touchable = Touchable.disabled
            switchButton.color = Color.GRAY
        }

        guardianStatusWidget.initialize(guardian)
        val sprite = Services.Media().getMonsterSprite(guardian.speciesID, guardian.currentForm)
        guardianImg.drawable = TextureRegionDrawable(sprite)
    }
}
